Importing the Libraries

In [112]:
import pandas as pd

Importing the Excel files

In [113]:
cust_df = pd.read_csv(r'C:\Users\Samyak\OneDrive\Documents\ASEP\EDI\dataset\raw\olist_customers_dataset.csv')
orders_df = pd.read_csv(r'C:\Users\Samyak\OneDrive\Documents\ASEP\EDI\dataset\raw\olist_orders_dataset.csv')
products_df = pd.read_csv(r'C:\Users\Samyak\OneDrive\Documents\ASEP\EDI\dataset\raw\olist_products_dataset.csv')
order_item_df = pd.read_csv(r'C:\Users\Samyak\OneDrive\Documents\ASEP\EDI\dataset\raw\olist_order_items_dataset.csv')
order_payment_df = pd.read_csv(r'C:\Users\Samyak\OneDrive\Documents\ASEP\EDI\dataset\raw\olist_order_payments_dataset.csv')
order_review_df = pd.read_csv(r'C:\Users\Samyak\OneDrive\Documents\ASEP\EDI\dataset\raw\olist_order_reviews_dataset.csv')

cust_df.head()





,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


Cleaning of Customer dataset

Run drop columns whenever you are running entire code block



In [114]:
cust_df.info()

cust_df = cust_df.drop(columns=['customer_zip_code_prefix'])
cust_df['customer_unique_id'].nunique()

cust_df.info()
cust_df['customer_unique_id'].nunique()




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         99441 non-null  object
 1   customer_unique_id  99441 non-null  object
 2   customer_city       99441 non-null  object
 3   customer_state      99441 non-null  object
dtypes: object(4)
memory usage: 3.0+ MB


96096

Cleaning of the Order_items dataset

In [115]:
order_item_df.info()

# order_item_df = order_item_df.drop(columns=['seller_id','shipping_limit_date'])


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


Cleaning of the Order_payment dataset


In [116]:
order_payment_df.info()

order_payment_df['order_id'].value_counts()

# order_payment_df = order_payment_df.drop(columns=['payment_sequential','payment_installments'])

order_payment_df= (
    order_payment_df
    .groupby("order_id")
    .agg(
        payment_value=("payment_value", "sum"),
        payment_type=("payment_type",
                      lambda x: x.iloc[0] if x.nunique() == 1 else "Multiple")
    )
    .reset_index()
)

order_payment_df.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


,order_id,payment_value,payment_type
0,00010242fe8c5a6d1ba2dd792cb16214,72.19,credit_card
1,00018f77f2f0320c557190d7a144bdd3,259.83,credit_card
2,000229ec398224ef6ca0657da4fc703e,216.87,credit_card
3,00024acbcdf0a6daa1e931b038114c75,25.78,credit_card
4,00042b26cf59d7ce69dfabb4e55b4fd9,218.04,credit_card


Cleaning of the Orders_review Dataset


In [117]:
# order_review_df = order_review_df.drop(columns=['review_comment_title','review_comment_message','review_creation_date','review_answer_timestamp'])

order_review_df.info()

order_review_df = (order_review_df.groupby('order_id',as_index=False).agg({'review_score':'mean'}))

order_review_df.head()
order_review_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 98673 entries, 0 to 98672
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   order_id      98673 non-null  object 
 1   review_score  98673 non-null  float64
dtypes: float64(1), object(1)
memory usage: 1.5+ MB


Cleaning of the Orders Dataset

In [118]:
orders_df.info()

# orders_df = orders_df.drop(columns=['order_approved_at','order_delivered_carrier_date','order_delivered_customer_date','order_estimated_delivery_date'])

orders_df['order_id'].value_counts()
orders_df['customer_id'].value_counts()


orders_df
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       

Cleaning the Product Dataset 

In [119]:


# products_df = products_df.drop(columns=['product_name_lenght','product_description_lenght','product_photos_qty','product_weight_g','product_length_cm','product_height_cm','product_width_cm'])



products_df['product_category_name'] = products_df['product_category_name'].fillna("Other")

products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32951 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


Merging the order_item and Product dataset

In [120]:
order_items_grouped_df = (
    order_item_df.merge(products_df[["product_id", "product_category_name"]], on="product_id").groupby("order_id", as_index=False).agg
    (
        total_items=("order_item_id", "count"),
        total_price=("price", "sum"),
        total_freight=("freight_value", "sum"),
        product_ids=("product_id", lambda x: ",".join(x)),
        categories=("product_category_name", lambda x: ",".join(sorted(set(x))))
    )
)

In [121]:
order_items_grouped_df.tail(20)

,order_id,total_items,total_price,total_freight,product_ids,categories
98646,fff2cdc825f9fc0ba3c04227cfa02303,1,24.99,25.63,044f05bc9de36e8a693a83e4bc79dd0d,telefonia
98647,fff2e9e3aa8644e19710216b4ef53ab2,1,69.90,16.25,7340a3839a1de1e99d149b8cf052a2ec,cama_mesa_banho
98648,fff3983dfa3c5a0d752d8d17baa406a0,1,66.39,14.05,092be1e8336fc404c57bd5970d056886,alimentos
98649,fff60e5408a9dd1e92ee30023052af30,1,129.90,18.80,1a405418406359cc2b8815f93bf359c2,esporte_lazer
98650,fff6889749958e42b47a7977a4cf0ea0,1,92.00,31.60,75f6a4f019ec1322758d53b2fee2cc12,malas_acessorios
98651,fff6b8ca971f8e3ec822e99d0f2d3d21,1,199.00,16.83,9afaad66aca8b0c79e4f084a89c9c92b,telefonia
98652,fff7c4452f050315db1b3f24d9df5fcd,1,736.00,20.91,dd469c03ad67e201bc2179ef077dcd48,relogios_presentes
98653,fff8286f77788ab8b55b2e5747fa7dd8,1,24.99,15.28,a2da86fa759178e9e58e54aa1a144e59,telefonia
98654,fff8287bbae429a99bb7e8c21d151c41,2,360.00,96.28,"bee2e070c39f3dd2f6883a17a5f0da45,bee2e070c39f3...",informatica_acessorios
98655,fff90cdcb3b2e6cfb397d05d562fd3fe,1,89.90,11.83,764292b2b0f73f77a0272be03fdd45f3,moveis_decoracao


Merging all the required columns into one table

In [122]:
master_df = (
    orders_df
    .merge(cust_df, on="customer_id", how="left")
    .merge(order_items_grouped_df, on="order_id", how="left")
    .merge(order_payment_df, on="order_id", how="left")
    .merge(order_review_df, on="order_id", how="left")
)

In [123]:
master_df = master_df.drop(columns=[
    "order_approved_at",'order_delivered_carrier_date','order_delivered_customer_date','order_estimated_delivery_date'

    ])

In [124]:
master_df[['order_date',"order_time"]] = master_df['order_purchase_timestamp'].str.split(' ',expand=True)


In [125]:
master_df.shape

(99441, 17)

In [126]:
master_df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,customer_unique_id,customer_city,customer_state,total_items,total_price,total_freight,product_ids,categories,payment_value,payment_type,review_score,order_date,order_time
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,7c396fd4830fd04220f754e42b4e5bff,sao paulo,SP,1.0,29.99,8.72,87285b34884572647811a353c7ac498a,utilidades_domesticas,38.71,Multiple,4.0,2017-10-02,10:56:33
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,af07308b275d755c9edb36a90c618231,barreiras,BA,1.0,118.70,22.76,595fac2a385ac33a80bd5114aec74eb8,perfumaria,141.46,boleto,4.0,2018-07-24,20:41:37
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,3a653a41f6f9fc3d2a113cf8398680e8,vianopolis,GO,1.0,159.90,19.22,aa4383b373c6aca5d8797843e5594415,automotivo,179.12,credit_card,5.0,2018-08-08,08:38:49
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,7c142cf63193a1473d2e66489a9ae977,sao goncalo do amarante,RN,1.0,45.00,27.20,d0b61bfb1de832b15ba9d266ca96e5b0,pet_shop,72.20,credit_card,5.0,2017-11-18,19:28:06
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,72632f0f9dd73dfee390c9b22eb56dd6,santo andre,SP,1.0,19.90,8.72,65266b2da20d04dbe00c5c2d3bb7859e,papelaria,28.62,credit_card,5.0,2018-02-13,21:18:39


In [127]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 17 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   order_id                  99441 non-null  object 
 1   customer_id               99441 non-null  object 
 2   order_status              99441 non-null  object 
 3   order_purchase_timestamp  99441 non-null  object 
 4   customer_unique_id        99441 non-null  object 
 5   customer_city             99441 non-null  object 
 6   customer_state            99441 non-null  object 
 7   total_items               98666 non-null  float64
 8   total_price               98666 non-null  float64
 9   total_freight             98666 non-null  float64
 10  product_ids               98666 non-null  object 
 11  categories                98666 non-null  object 
 12  payment_value             99440 non-null  float64
 13  payment_type              99440 non-null  object 
 14  review

In [128]:
master_df['payment_type'] = master_df['payment_type'].fillna('credit_card')

In [129]:
payment = master_df['total_price'] + master_df['total_freight']

master_df['payment_value'] = master_df['payment_value'].fillna(payment)

In [130]:
master_df['review_score'] = master_df['review_score'].fillna(0)

In [131]:
master_df["categories"] = master_df["categories"].fillna("Other")

In [132]:
invalid_orders = master_df[master_df["total_items"].isna()].copy()

master_df = master_df.dropna(subset=["total_items"]).reset_index(drop=True)

In [133]:
print(master_df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 98666 entries, 0 to 98665
Data columns (total 17 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   order_id                  98666 non-null  object 
 1   customer_id               98666 non-null  object 
 2   order_status              98666 non-null  object 
 3   order_purchase_timestamp  98666 non-null  object 
 4   customer_unique_id        98666 non-null  object 
 5   customer_city             98666 non-null  object 
 6   customer_state            98666 non-null  object 
 7   total_items               98666 non-null  float64
 8   total_price               98666 non-null  float64
 9   total_freight             98666 non-null  float64
 10  product_ids               98666 non-null  object 
 11  categories                98666 non-null  object 
 12  payment_value             98666 non-null  float64
 13  payment_type              98666 non-null  object 
 14  review

# Summary :
1. All the tables which were having null values and duplicate values are grouped and handled the null values
2. Merging multiple tables into 1 single large table which we will use for the feature engineering and further.
3. The master table contains  98666 records 
4. In master table some columns are kept for future reference like order_purchase_timestamp and order_time

# Feature Engineering


# For Sales Agent

1. Month ---> order_date
2. Quarter --> order_Date
3. day_of_week --> order_Date
4. year --> order_date

In [134]:
master_df["order_date"] = pd.to_datetime(master_df["order_date"])
master_df['order_year'] = master_df['order_date'].dt.year
master_df['order_month'] = master_df['order_date'].dt.month_name()
master_df['order_quarter'] = master_df['order_date'].dt.quarter
master_df['order_day_of_week'] = master_df['order_date'].dt.day_name()

In [135]:
master_df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,customer_unique_id,customer_city,customer_state,total_items,total_price,total_freight,...,categories,payment_value,payment_type,review_score,order_date,order_time,order_year,order_month,order_quarter,order_day_of_week
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,7c396fd4830fd04220f754e42b4e5bff,sao paulo,SP,1.0,29.99,8.72,...,utilidades_domesticas,38.71,Multiple,4.0,2017-10-02,10:56:33,2017,October,4,Monday
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,af07308b275d755c9edb36a90c618231,barreiras,BA,1.0,118.70,22.76,...,perfumaria,141.46,boleto,4.0,2018-07-24,20:41:37,2018,July,3,Tuesday
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,3a653a41f6f9fc3d2a113cf8398680e8,vianopolis,GO,1.0,159.90,19.22,...,automotivo,179.12,credit_card,5.0,2018-08-08,08:38:49,2018,August,3,Wednesday
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,7c142cf63193a1473d2e66489a9ae977,sao goncalo do amarante,RN,1.0,45.00,27.20,...,pet_shop,72.20,credit_card,5.0,2017-11-18,19:28:06,2017,November,4,Saturday
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,72632f0f9dd73dfee390c9b22eb56dd6,santo andre,SP,1.0,19.90,8.72,...,papelaria,28.62,credit_card,5.0,2018-02-13,21:18:39,2018,February,1,Tuesday


In [136]:
monthly_revenue_df = (master_df.groupby(['order_year','order_month'],as_index=False).agg(
    monthly_revenue = ('payment_value','sum')
))

monthly_orders_df = (master_df.groupby(['order_month','order_year'],as_index=False).agg(
    monthly_orders = ('order_id','count')
))

sales_df = monthly_revenue_df.merge(
    monthly_orders_df, on = ['order_year','order_month']
)

In [137]:
sales_df

,order_year,order_month,monthly_revenue,monthly_orders
0,2016,December,19.62,1
1,2016,October,56884.89,308
2,2016,September,354.75,3
3,2017,April,412562.01,2391
4,2017,August,668372.85,4293
5,2017,December,863675.00,5624
6,2017,February,286340.87,1733
7,2017,January,137251.79,789
8,2017,July,585076.47,3969
9,2017,June,503138.27,3217


In [138]:
sales_df["holiday_flag"] = sales_df["order_month"].isin(['January', 'April', 'August', 'October', 'November', 'December']).astype(int)

In [139]:
sales_df["festival_flag"] = sales_df["order_month"].isin(['August', 'October', 'November']).astype(int)

In [140]:
import numpy as np

sales_df["discount_percentage"] = 0

# Festival Months (High Discount)

festival_mask = sales_df["festival_flag"] == 1

sales_df.loc[festival_mask, "discount_percentage"] = np.random.choice(
    [10, 15, 20, 25, 30],
    size=festival_mask.sum(),
    p=[0.15, 0.25, 0.30, 0.20, 0.10]
)

# Holiday Months (Non-Festival means less discount)
holiday_mask = (
    (sales_df["holiday_flag"] == 1) &
    (sales_df["festival_flag"] == 0)
)

sales_df.loc[holiday_mask, "discount_percentage"] = np.random.choice(
    [5, 10, 15],
    size=holiday_mask.sum(),
    p=[0.50, 0.35, 0.15]
)

# Normal Months (No discount)
normal_mask = (
    (sales_df["holiday_flag"] == 0) &
    (sales_df["festival_flag"] == 0)
)

sales_df.loc[normal_mask, "discount_percentage"] = np.random.choice(
    [0, 5, 10],
    size=normal_mask.sum(),
    p=[0.60, 0.30, 0.10]
)



In [141]:
sales_df.head()

,order_year,order_month,monthly_revenue,monthly_orders,holiday_flag,festival_flag,discount_percentage
0,2016,December,19.62,1,1,0,5
1,2016,October,56884.89,308,1,1,10
2,2016,September,354.75,3,0,0,0
3,2017,April,412562.01,2391,1,0,15
4,2017,August,668372.85,4293,1,1,10


# Summary for Sales Agent

1. Mostly all order details in the master_df
2. All sales regarding information in the sales_df
3. Target columns will be monthly_revenue

# Customer Agent
Grouping with the customer_unique_id
1. total_orders --> order_id 
2. total_spend  --> payment_value
3. avg_order_value --> payment_value (mean)
4. recency_days -->order_date
5. purchase_frequency --> total_order / active months
6. avg_days_between_orders  --> order_date

In [142]:
master_df = master_df.sort_values(by=['customer_unique_id','order_date'])

In [143]:
refrence_date = master_df['order_date'].max()

In [144]:
customer_features = (master_df.groupby('customer_unique_id',as_index=False).agg(
    customer_state = ('customer_state','first'),
    customer_city = ('customer_city','first'),
    total_orders = ('order_id','count'),
    total_spend = ('payment_value','sum'),
    avg_order_value = ('payment_value','mean'),
    first_purchase = ('order_date','min'),
    last_purchase = ('order_date','max'),
    avg_review_score = ('review_score','mean')

))

In [145]:
customer_features['recency_days'] = (refrence_date - customer_features['last_purchase']).dt.days

In [146]:
customer_features['customer_active_days'] = (customer_features['last_purchase'] - customer_features['first_purchase']).dt.days + 1

In [147]:
customer_features['purchase_frequency'] =( customer_features['total_orders']/(customer_features['customer_active_days']/30))

In [148]:
import numpy as np

# Replace any infinite values

customer_features['purchase_frequency'] = (customer_features['purchase_frequency'].replace([np.inf , -np.inf],0).fillna(0))

In [149]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 98666 entries, 52389 to 69474
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   order_id                  98666 non-null  object        
 1   customer_id               98666 non-null  object        
 2   order_status              98666 non-null  object        
 3   order_purchase_timestamp  98666 non-null  object        
 4   customer_unique_id        98666 non-null  object        
 5   customer_city             98666 non-null  object        
 6   customer_state            98666 non-null  object        
 7   total_items               98666 non-null  float64       
 8   total_price               98666 non-null  float64       
 9   total_freight             98666 non-null  float64       
 10  product_ids               98666 non-null  object        
 11  categories                98666 non-null  object        
 12  payment_value      

In [150]:
customer_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95420 entries, 0 to 95419
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   customer_unique_id    95420 non-null  object        
 1   customer_state        95420 non-null  object        
 2   customer_city         95420 non-null  object        
 3   total_orders          95420 non-null  int64         
 4   total_spend           95420 non-null  float64       
 5   avg_order_value       95420 non-null  float64       
 6   first_purchase        95420 non-null  datetime64[ns]
 7   last_purchase         95420 non-null  datetime64[ns]
 8   avg_review_score      95420 non-null  float64       
 9   recency_days          95420 non-null  int64         
 10  customer_active_days  95420 non-null  int64         
 11  purchase_frequency    95420 non-null  float64       
dtypes: datetime64[ns](2), float64(4), int64(3), object(3)
memory usage: 8.7+ M

In [151]:
avg_gap = (
    master_df
    .sort_values(["customer_unique_id", "order_date"])
)

avg_gap["gap"] = (
    avg_gap
    .groupby("customer_unique_id")["order_date"]
    .diff()
    .dt.days
)

gap_df = (
    avg_gap
    .groupby("customer_unique_id", as_index=False)
    .agg(
        avg_days_between_orders=("gap", "mean")
    )
)

In [152]:
gap_df["avg_days_between_orders"] = gap_df["avg_days_between_orders"].fillna(0)

In [153]:
customer_features = customer_features.merge(gap_df , on='customer_unique_id',how='left')

In [154]:
customer_features = customer_features.drop(
    columns=[
        "first_purchase",
        "last_purchase",
        "customer_active_days"
    ]
)

In [155]:
customer_features['churn'] = (customer_features['recency_days']>90).astype(int)

In [156]:
customer_features.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95420 entries, 0 to 95419
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer_unique_id       95420 non-null  object 
 1   customer_state           95420 non-null  object 
 2   customer_city            95420 non-null  object 
 3   total_orders             95420 non-null  int64  
 4   total_spend              95420 non-null  float64
 5   avg_order_value          95420 non-null  float64
 6   avg_review_score         95420 non-null  float64
 7   recency_days             95420 non-null  int64  
 8   purchase_frequency       95420 non-null  float64
 9   avg_days_between_orders  95420 non-null  float64
 10  churn                    95420 non-null  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 8.0+ MB


# Inventory Agent 
group by product_id
1. historical_sales --> count of product_id
2. avg_daily_sales  --> historical_sales + order_date
3. monthly_sales --> order_month

In [157]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 98666 entries, 52389 to 69474
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   order_id                  98666 non-null  object        
 1   customer_id               98666 non-null  object        
 2   order_status              98666 non-null  object        
 3   order_purchase_timestamp  98666 non-null  object        
 4   customer_unique_id        98666 non-null  object        
 5   customer_city             98666 non-null  object        
 6   customer_state            98666 non-null  object        
 7   total_items               98666 non-null  float64       
 8   total_price               98666 non-null  float64       
 9   total_freight             98666 non-null  float64       
 10  product_ids               98666 non-null  object        
 11  categories                98666 non-null  object        
 12  payment_value      

In [158]:
inventory_df = order_item_df.merge(
    products_df[['product_id' , 'product_category_name']],
    on = 'product_id',
    how= 'left'
)

In [159]:
inventory_df = inventory_df.drop(columns="shipping_limit_date")

In [160]:
inventory_df = inventory_df.merge(
    orders_df[['order_id','order_purchase_timestamp','order_status']],
    on = 'order_id',
    how ='left'
)

In [161]:
inventory_df = inventory_df[inventory_df['order_status']=='delivered'].reset_index(drop=True)

In [162]:
inventory_df[['order_date','order_time']] = inventory_df['order_purchase_timestamp'].str.split(' ',expand=True)

In [163]:
inventory_df['order_date'] = pd.to_datetime(inventory_df['order_date'])

In [164]:
historical = (inventory_df.groupby('product_id').size().reset_index(name='historical_sales'))

In [165]:
inventory_df['order_year'] = inventory_df['order_date'].dt.year
inventory_df['order_month'] = inventory_df['order_date'].dt.month

monthly_sales_df = (inventory_df.groupby(['product_id','order_year','order_month'], as_index=False).agg(
    monthly_sales = ('order_item_id','count')
))

In [166]:
daily_sales = (
    inventory_df
    .groupby(
        ["product_id", "order_date"],
        as_index=False
    )
    .agg(
        daily_sales=("order_item_id", "count")
    )
)

avg_daily_sales = (
    daily_sales
    .groupby("product_id", as_index=False)
    .agg(
        avg_daily_sales=("daily_sales", "mean")
    )
)

In [167]:
inventory_df = (
    historical
    .merge(avg_daily_sales, on="product_id", how="left")
    .merge(monthly_sales_df, on="product_id", how="left")
)

In [168]:
inventory_df

,product_id,historical_sales,avg_daily_sales,order_year,order_month,monthly_sales
0,00066f42aeeb9f3007548bb9d3f33c38,1,1.0,2018,5,1
1,00088930e925c41fd95ebfe695fd2655,1,1.0,2017,12,1
2,0009406fd7479715e4bef61dd91f2462,1,1.0,2017,12,1
3,000b8f95fcb9e0096488278317764d19,2,1.0,2018,8,2
4,000d9be29b5207b54e86aa1b1ac54872,1,1.0,2018,4,1
...,...,...,...,...,...,...
60876,fff9553ac224cec9d15d49f5a263411f,1,1.0,2017,10,1
60877,fffdb2d0ec8d6a61f0a0a0db3f25b441,5,1.0,2018,4,2
60878,fffdb2d0ec8d6a61f0a0a0db3f25b441,5,1.0,2018,5,1
60879,fffdb2d0ec8d6a61f0a0a0db3f25b441,5,1.0,2018,8,2


In [169]:
# Product with higher historical sales usually has more stock

inventory_df["current_stock"] = (
    inventory_df["historical_sales"] *
    np.random.uniform(0.25, 0.60, len(inventory_df))
).round().astype(int)

inventory_df["current_stock"] = (
    inventory_df["current_stock"]
    .clip(lower=10)
)

In [170]:
inventory_df["reorder_level"] = (
    inventory_df["current_stock"] *
    np.random.uniform(0.30, 0.40, len(inventory_df))
).round().astype(int)

In [171]:
inventory_df['supplier_id'] = order_item_df['seller_id']

In [172]:

seller_ids = inventory_df["supplier_id"].unique()


lead_time_map = {
    seller: np.random.randint(2, 11)   # 2–10 days
    for seller in seller_ids
}

inventory_df["lead_time_days"] = (
    inventory_df["supplier_id"]
    .map(lead_time_map)
)

In [173]:
inventory_df["reorder_level"] = (
    inventory_df["current_stock"] *
    np.random.uniform(0.30, 0.40, len(inventory_df))
).round().astype(int)

In [174]:
inventory_df["reorder_quantity"] = np.where(
    inventory_df["current_stock"] <= inventory_df["reorder_level"],
    inventory_df["monthly_sales"],0
)

In [175]:
warehouse_list = [
    "Mumbai",
    "Delhi",
    "Bangalore",
    "Hyderabad",
    "Pune",
    "Chennai",
    "Ahmedabad",
    "Indore",
    
]

inventory_df["warehouse_location"] = np.random.choice(
    warehouse_list,
    size=len(inventory_df),
    p=[0.22 , 0.13 ,0.10 ,0.10 ,0.15, 0.11 ,0.12, 0.07]
)

In [176]:
inventory_df['category'] = products_df['product_category_name']

In [177]:
inventory_df['category'] = inventory_df['category'].fillna('Other')

In [178]:
inventory_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60881 entries, 0 to 60880
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   product_id          60881 non-null  object 
 1   historical_sales    60881 non-null  int64  
 2   avg_daily_sales     60881 non-null  float64
 3   order_year          60881 non-null  int32  
 4   order_month         60881 non-null  int32  
 5   monthly_sales       60881 non-null  int64  
 6   current_stock       60881 non-null  int64  
 7   reorder_level       60881 non-null  int64  
 8   supplier_id         60881 non-null  object 
 9   lead_time_days      60881 non-null  int64  
 10  reorder_quantity    60881 non-null  int64  
 11  warehouse_location  60881 non-null  object 
 12  category            60881 non-null  object 
dtypes: float64(1), int32(2), int64(6), object(4)
memory usage: 5.6+ MB


# Finance Agent


In [179]:
finance_df = master_df[[
    'order_id',
    'order_date',
    'payment_value',
    'total_items',
    'total_freight'
]].copy()

In [180]:
finance_df['order_date'] = pd.to_datetime(finance_df['order_date'])

In [181]:
finance_df

,order_id,order_date,payment_value,total_items,total_freight
52389,e22acc9c116caa3f2b7121bbb380d08e,2018-05-10,141.90,1.0,12.00
73317,3594e05a005ac4d06a72673270ef9ec9,2018-05-07,27.19,1.0,8.29
26249,b33ec3b699337181488304f362a6b734,2017-03-10,86.22,1.0,17.22
97729,41272756ecddd9a9ed0180413cc22fb6,2017-10-12,43.62,1.0,17.63
41239,d957021f1127559cd947b62533f484f7,2017-11-14,196.89,1.0,16.89
...,...,...,...,...,...
86324,725cf8e9c24e679a8a5a32cb92c9ce1e,2017-06-08,2067.42,2.0,497.42
20396,c71b9252fd7b3b263aaa4cb09319a323,2017-12-10,84.58,1.0,19.69
43602,fdc45e6c7555e6cb3cc0daca2557dbe1,2017-02-07,112.46,1.0,22.56
96105,94d3ee0bc2a0af9d4fa47a4d63616e8d,2018-05-02,133.69,1.0,18.69


In [182]:
finance_df["order_year"] = finance_df["order_date"].dt.year

finance_df["order_month"] = finance_df["order_date"].dt.month

finance_df["order_quarter"] = finance_df["order_date"].dt.quarter

In [183]:
master_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 98666 entries, 52389 to 69474
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   order_id                  98666 non-null  object        
 1   customer_id               98666 non-null  object        
 2   order_status              98666 non-null  object        
 3   order_purchase_timestamp  98666 non-null  object        
 4   customer_unique_id        98666 non-null  object        
 5   customer_city             98666 non-null  object        
 6   customer_state            98666 non-null  object        
 7   total_items               98666 non-null  float64       
 8   total_price               98666 non-null  float64       
 9   total_freight             98666 non-null  float64       
 10  product_ids               98666 non-null  object        
 11  categories                98666 non-null  object        
 12  payment_value      

In [184]:
finance_df = (
    finance_df
    .groupby(
        ["order_year", "order_month", "order_quarter"],
        as_index=False
    )
    .agg(
        monthly_revenue=("payment_value", "sum"),
        total_orders=("order_id", "count"),
        total_qty_sold = ('total_items','sum'),
        total_freight_cost = ('total_freight','sum')
    )
)

In [185]:
import numpy as np

np.random.seed(42)

finance_df["purchase_cost"] = (
    finance_df["monthly_revenue"] *
    np.random.uniform(0.55, 0.75, len(finance_df))
).round(2)

In [186]:
finance_df["marketing_cost"] = (
    finance_df["monthly_revenue"] *
    np.random.uniform(0.03, 0.08, len(finance_df))
).round(2)

In [187]:
finance_df["salary_cost"] = np.random.randint(
    50000,
    100001,
    len(finance_df)
)

In [188]:
finance_df["rent_cost"] = np.random.randint(
    20000,
    50001,
    len(finance_df)
)

In [189]:
finance_df["gross_profit"] = (
    finance_df["monthly_revenue"]
    - finance_df["purchase_cost"]
    -finance_df['total_freight_cost']
)

# Gross Profit = S.P - C.P 

In [190]:
finance_df["net_profit"] = (
    finance_df["monthly_revenue"]
    - finance_df["purchase_cost"]
    - finance_df["marketing_cost"]
    - finance_df["salary_cost"]
    - finance_df["rent_cost"]
    -finance_df['total_freight_cost']
)

In [191]:
finance_df['profit_margin'] = (finance_df["net_profit"] / finance_df["monthly_revenue"]) *100

In [192]:
finance_df

,order_year,order_month,order_quarter,monthly_revenue,total_orders,total_qty_sold,total_freight_cost,purchase_cost,marketing_cost,salary_cost,rent_cost,gross_profit,net_profit,profit_margin
0,2016,9,3,354.75,3,6.0,87.39,221.69,18.73,63417,32185,45.67,-95575.06,-26941.525018
1,2016,10,4,56884.89,308,363.0,7301.18,42102.95,3939.78,90757,41243,7480.76,-128459.02,-225.822745
2,2016,12,4,19.62,1,1.0,8.72,13.66,0.78,59692,26331,-2.76,-86026.54,-438463.506626
3,2017,1,1,137251.79,789,955.0,16875.62,91921.87,7646.53,95758,28571,28454.30,-103521.23,-75.424321
4,2017,2,1,286340.87,1733,1951.0,38977.60,166422.38,17071.85,56873,45068,80940.89,-38071.96,-13.296027
5,2017,3,1,432087.03,2641,3000.0,57704.29,251128.51,13966.14,55675,27208,123254.23,26405.09,6.111058
6,2017,4,2,412562.01,2391,2684.0,52495.01,231701.72,24909.36,50161,25276,128365.28,28018.92,6.791445
7,2017,5,2,586406.28,3660,4136.0,80119.81,424109.68,22592.01,87065,38446,82176.79,-65926.22,-11.242414
8,2017,6,2,503138.27,3217,3583.0,69924.44,337214.84,16730.65,76557,48460,95998.99,-45748.66,-9.092662
9,2017,7,3,585076.47,3969,4519.0,86940.14,404647.38,45310.82,83763,36448,93488.95,-72032.87,-12.311702


# Summary after Feature Engineering

All the tables for the Agent are created

Sales Agent --> sales_df

Customer Agent --> customer_features

Inventory Agent --> inventory_df

Finance Agent --> finance_df

In [193]:
print(master_df.info())
print("\n ")
print(sales_df.info())
print("\n ")
print(customer_features.info())
print("\n ")
print(inventory_df.info())
print("\n ")
print(finance_df.info())
print("\n ")

<class 'pandas.core.frame.DataFrame'>
Index: 98666 entries, 52389 to 69474
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   order_id                  98666 non-null  object        
 1   customer_id               98666 non-null  object        
 2   order_status              98666 non-null  object        
 3   order_purchase_timestamp  98666 non-null  object        
 4   customer_unique_id        98666 non-null  object        
 5   customer_city             98666 non-null  object        
 6   customer_state            98666 non-null  object        
 7   total_items               98666 non-null  float64       
 8   total_price               98666 non-null  float64       
 9   total_freight             98666 non-null  float64       
 10  product_ids               98666 non-null  object        
 11  categories                98666 non-null  object        
 12  payment_value      

In [194]:
master_df.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'customer_unique_id', 'customer_city', 'customer_state', 'total_items',
       'total_price', 'total_freight', 'product_ids', 'categories',
       'payment_value', 'payment_type', 'review_score', 'order_date',
       'order_time', 'order_year', 'order_month', 'order_quarter',
       'order_day_of_week'],
      dtype='object')

In [195]:
master_df["product_count"] = (
    master_df["product_ids"]
    .fillna("")
    .apply(lambda x: len(str(x).split(",")) if x != "" else 0)
)

In [196]:
master_df = master_df.drop(columns=['order_purchase_timestamp','product_ids'])

In [197]:
master_df.rename(
    columns={
        "categories": "product_categories"
    },
    inplace=True
)

In [198]:
master_df.columns

Index(['order_id', 'customer_id', 'order_status', 'customer_unique_id',
       'customer_city', 'customer_state', 'total_items', 'total_price',
       'total_freight', 'product_categories', 'payment_value', 'payment_type',
       'review_score', 'order_date', 'order_time', 'order_year', 'order_month',
       'order_quarter', 'order_day_of_week', 'product_count'],
      dtype='object')

In [199]:
master_df = master_df[
    [
        "order_id",
        "customer_id",
        "customer_unique_id",
        "customer_city",
        "customer_state",
        "order_status",
        "order_date",
        "order_time",
        "order_year",
        "order_month",
        "order_quarter",
        "order_day_of_week",
        "product_count",
        "product_categories",
        "total_items",
        "total_price",
        "total_freight",
        "payment_value",
        "payment_type",
        "review_score"
    ]
]

In [200]:
sales_df = sales_df[
    [
        "order_year",
        "order_month",
        "monthly_revenue",
        "monthly_orders",
        "holiday_flag",
        "festival_flag",
        "discount_percentage"
    ]
]

In [201]:
inventory_df = inventory_df[
    [
        "product_id",
        "category",
        "supplier_id",
        "order_year",
        "order_month",
        "historical_sales",
        "avg_daily_sales",
        "monthly_sales",
        "current_stock",
        "reorder_level",
        "lead_time_days",
        "reorder_quantity",
        "warehouse_location"
    ]
]

In [202]:
customer_df = customer_features[
    [
        "customer_unique_id",
        "customer_state",
        "customer_city",
        "total_orders",
        "total_spend",
        "avg_order_value",
        "avg_review_score",
        "recency_days",
        "purchase_frequency",
        "avg_days_between_orders",
        "churn"
    ]
]

In [203]:
finance_df = finance_df[
    [
        "order_year",
        "order_month",
        "order_quarter",
        "monthly_revenue",
        "total_orders",
        "total_qty_sold",
        "total_freight_cost",
        "purchase_cost",
        "marketing_cost",
        "salary_cost",
        "rent_cost",
        "gross_profit",
        "net_profit",
        "profit_margin"
    ]
]

In [204]:
product_df = (
    order_item_df
    .sort_values(["order_id", "order_item_id"])
    .groupby("order_id", as_index=False)
    .first()[["order_id", "product_id"]]
)

In [205]:
master_df = master_df.merge(
    product_df,
    on="order_id",
    how="left"
)

In [206]:
master_df.head()

,order_id,customer_id,customer_unique_id,customer_city,customer_state,order_status,order_date,order_time,order_year,order_month,...,order_day_of_week,product_count,product_categories,total_items,total_price,total_freight,payment_value,payment_type,review_score,product_id
0,e22acc9c116caa3f2b7121bbb380d08e,fadbb3709178fc513abc1b2670aa1ad2,0000366f3b9a7992bf8c76cfdf3221e2,cajamar,SP,delivered,2018-05-10,10:56:27,2018,May,...,Thursday,1,cama_mesa_banho,1.0,129.90,12.00,141.90,credit_card,5.0,372645c7439f9661fbbacfd129aa92ec
1,3594e05a005ac4d06a72673270ef9ec9,4cb282e167ae9234755102258dd52ee8,0000b849f77a49e4a4ce2b2a4ca5be3f,osasco,SP,delivered,2018-05-07,11:11:27,2018,May,...,Monday,1,beleza_saude,1.0,18.90,8.29,27.19,credit_card,4.0,5099f7000472b634fea8304448d20825
2,b33ec3b699337181488304f362a6b734,9b3932a6253894a02c1df9d19004239f,0000f46a3911fa3c0805444483337064,sao jose,SC,delivered,2017-03-10,21:05:03,2017,March,...,Friday,1,papelaria,1.0,69.00,17.22,86.22,credit_card,3.0,64b488de448a5324c4134ea39c28a34b
3,41272756ecddd9a9ed0180413cc22fb6,914991f0c02ef0843c0e7010c819d642,0000f6ccb0745a6a4b88665a16c9f078,belem,PA,delivered,2017-10-12,20:29:41,2017,October,...,Thursday,1,telefonia,1.0,25.99,17.63,43.62,credit_card,4.0,2345a354a6f2033609bbf62bf5be9ef6
4,d957021f1127559cd947b62533f484f7,47227568b10f5f58a524a75507e6992c,0004aac84e0df4da2b147fca70cf8255,sorocaba,SP,delivered,2017-11-14,19:45:42,2017,November,...,Tuesday,1,telefonia,1.0,180.00,16.89,196.89,credit_card,5.0,c72e18b3fe2739b8d24ebf3102450f37


In [207]:
master_df = master_df[
    [
        "order_id",
        "customer_id",
        "customer_unique_id",
        "customer_city",
        "customer_state",
        "product_id",
        "product_count",
        "product_categories",
        "order_status",
        "order_date",
        "order_time",
        "order_year",
        "order_month",
        "order_quarter",
        "order_day_of_week",

        "total_items",
        "total_price",
        "total_freight",
        "payment_value",
        "payment_type",
        "review_score"
    ]
]

In [208]:
import numpy as np

# For reproducibility
np.random.seed(42)

# Get unique customers
unique_customers = master_df["customer_unique_id"].unique()

# Generate unique Indian mobile numbers
phone_numbers = np.random.choice(
    np.arange(9000000000, 9999999999),
    size=len(unique_customers),
    replace=False
)

# Create mapping
phone_map = dict(zip(unique_customers, phone_numbers))

# Add phone number column
master_df["customer_phone"] = (
    master_df["customer_unique_id"]
    .map(phone_map)
    .astype(str)
)

In [209]:
master_df = master_df[
    [
        "order_id",
        "customer_id",
        "customer_unique_id",
        "customer_phone",
        "customer_city",
        "customer_state",
        "product_id",
        "product_count",
        "product_categories",
        "order_status",
        "order_date",
        "order_time",
        "order_year",
        "order_month",
        "order_quarter",
        "order_day_of_week",
        "total_items",
        "total_price",
        "total_freight",
        "payment_value",
        "payment_type",
        "review_score"
    ]
]

In [210]:
import sqlite3
import pandas as pd

# Connect/Create Database
conn = sqlite3.connect("../database/business_ai.db")

# Push DataFrame to SQLite
master_df.to_sql(
    "transactions",
    conn,
    if_exists="replace",
    index=False
)





98666

In [211]:
import sqlite3
import pandas as pd

# Connect/Create Database
conn = sqlite3.connect("../database/business_ai.db")

# Push DataFrame to SQLite
finance_df.to_sql(
    "finance",
    conn,
    if_exists="replace",
    index=False
)

24

In [212]:
import sqlite3
import pandas as pd

# Connect/Create Database
conn = sqlite3.connect("../database/business_ai.db")

# Push DataFrame to SQLite
master_df.to_sql(
    "transactions",
    conn,
    if_exists="replace",
    index=False
)

98666

In [213]:
pd.read_sql("Select * from transactions",conn)

,order_id,customer_id,customer_unique_id,customer_phone,customer_city,customer_state,product_id,product_count,product_categories,order_status,...,order_year,order_month,order_quarter,order_day_of_week,total_items,total_price,total_freight,payment_value,payment_type,review_score
0,e22acc9c116caa3f2b7121bbb380d08e,fadbb3709178fc513abc1b2670aa1ad2,0000366f3b9a7992bf8c76cfdf3221e2,9821266650,cajamar,SP,372645c7439f9661fbbacfd129aa92ec,1,cama_mesa_banho,delivered,...,2018,May,2,Thursday,1.0,129.90,12.00,141.90,credit_card,5.0
1,3594e05a005ac4d06a72673270ef9ec9,4cb282e167ae9234755102258dd52ee8,0000b849f77a49e4a4ce2b2a4ca5be3f,9458682706,osasco,SP,5099f7000472b634fea8304448d20825,1,beleza_saude,delivered,...,2018,May,2,Monday,1.0,18.90,8.29,27.19,credit_card,4.0
2,b33ec3b699337181488304f362a6b734,9b3932a6253894a02c1df9d19004239f,0000f46a3911fa3c0805444483337064,9143269522,sao jose,SC,64b488de448a5324c4134ea39c28a34b,1,papelaria,delivered,...,2017,March,1,Friday,1.0,69.00,17.22,86.22,credit_card,3.0
3,41272756ecddd9a9ed0180413cc22fb6,914991f0c02ef0843c0e7010c819d642,0000f6ccb0745a6a4b88665a16c9f078,9035920466,belem,PA,2345a354a6f2033609bbf62bf5be9ef6,1,telefonia,delivered,...,2017,October,4,Thursday,1.0,25.99,17.63,43.62,credit_card,4.0
4,d957021f1127559cd947b62533f484f7,47227568b10f5f58a524a75507e6992c,0004aac84e0df4da2b147fca70cf8255,9174759669,sorocaba,SP,c72e18b3fe2739b8d24ebf3102450f37,1,telefonia,delivered,...,2017,November,4,Tuesday,1.0,180.00,16.89,196.89,credit_card,5.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
98661,725cf8e9c24e679a8a5a32cb92c9ce1e,74be082247cd677a147d83ee670e9d53,fffcf5a5ff07b0908bd4e2dbc735a684,9452243659,sanharo,PE,c100e5fef1abb5e1c5054d1dac2d83ac,2,beleza_saude,delivered,...,2017,June,2,Thursday,2.0,1570.00,497.42,2067.42,credit_card,5.0
98662,c71b9252fd7b3b263aaa4cb09319a323,0ecf8e0a08148af0bf313184f167670a,fffea47cd6d3cc0a88bd621562a9d061,9555028449,feira de santana,BA,251876e7ba013b87b7262fb2524e2d60,1,bebes,delivered,...,2017,December,4,Sunday,1.0,64.89,19.69,84.58,credit_card,4.0
98663,fdc45e6c7555e6cb3cc0daca2557dbe1,27f584b0f1dc4e610065c240f68b6be0,ffff371b4d645b6ecea244b27531430a,9523324544,sinop,MT,13ad6994bda56edfa0a269e3d9a9acb3,1,automotivo,delivered,...,2017,February,1,Tuesday,1.0,89.90,22.56,112.46,credit_card,5.0
98664,94d3ee0bc2a0af9d4fa47a4d63616e8d,832a3b0254347d409512ae92eaf154a6,ffff5962728ec6157033ef9805bacc48,9945606355,bom jesus do norte,ES,461f43be3bdf8844e65b62d9ac2c7a5a,1,relogios_presentes,delivered,...,2018,May,2,Wednesday,1.0,115.00,18.69,133.69,credit_card,5.0


In [214]:
import sqlite3

# Connect to the database
conn = sqlite3.connect("../database/business_ai.db")

cursor = conn.cursor()

# Create Product Inventory Table
cursor.execute("""
CREATE TABLE IF NOT EXISTS product_inventory (
    product_id TEXT PRIMARY KEY,
    product_name TEXT NOT NULL,
    category TEXT NOT NULL,
    current_stock INTEGER NOT NULL,
    selling_price REAL NOT NULL,
    purchase_price REAL NOT NULL,
    seller_id TEXT,
    warehouse_location TEXT

);
""")

conn.commit()

print("product_inventory table created successfully.")

conn.close()

product_inventory table created successfully.


In [215]:

conn = sqlite3.connect("../database/business_ai.db")

tables = pd.read_sql(
    "SELECT name FROM sqlite_master WHERE type='table';",
    conn
)

print(tables)

conn.close()

                name
0  product_inventory
1            finance
2       transactions


In [216]:
sales_df.to_csv(r'C:\Users\Samyak\OneDrive\Documents\ASEP\EDI\dataset\processed\sales_df.csv',index=False)

In [217]:
customer_df.to_csv(r'C:\Users\Samyak\OneDrive\Documents\ASEP\EDI\dataset\processed\customer_df.csv')
inventory_df.to_csv(r'C:\Users\Samyak\OneDrive\Documents\ASEP\EDI\dataset\processed\inventory_df.csv')
finance_df.to_csv(r'C:\Users\Samyak\OneDrive\Documents\ASEP\EDI\dataset\processed\finance_df.csv')

In [2]:
import sqlite3

conn = sqlite3.connect("../database/business_ai.db")
cursor = conn.cursor()

cursor.execute("""
CREATE TABLE IF NOT EXISTS users (
    user_id TEXT PRIMARY KEY,
    username TEXT NOT NULL UNIQUE,
    password TEXT NOT NULL,
    full_name TEXT NOT NULL,
    business_name TEXT NOT NULL,
    phone_number TEXT NOT NULL UNIQUE,
    email TEXT UNIQUE,
    role TEXT NOT NULL DEFAULT 'Admin'
);
""")

conn.commit()
conn.close()

print("Users table created successfully.")

Users table created successfully.
